diff --git a/contrib/zfs-fuse.automated.scrub.cron.d b/contrib/zfs-fuse.automated.scrub.cron.d
new file mode 100644
index 0000000..29f8d36
--- /dev/null
+++ b/contrib/zfs-fuse.automated.scrub.cron.d
@@ -0,0 +1,10 @@
+#
+# Copyright © martin f. krafft <madduck@madduck.net>
+# distributed under the terms of the Artistic Licence 2.0
+#
+
+# By default, run at 00:57 on every Sunday, but do nothing unless the day of
+# the month is less than or equal to 7. Thus, only run on the first Sunday of
+# each month. crontab(5) sucks, unfortunately, in this regard; therefore this
+# hack (see #380425).
+57 0 * * 0 root if [ $(date +\%d) -le 7 ]; then zpool list -H -o name | xargs zpool scrub; fi
-- 
1.7.10.4

